phper/src/builder/value.rs

```diff
@@ -0,0 +1,301 @@
+use crate::arrays::{InsertKey, ZArray};
+use crate::builder::describe::Describe;
+use crate::values::ZVal;
+use crate::Error;
+
+#[derive(Clone)]
+pub struct Value {
+    zval: ZVal,
+}
+
+impl Value {
+    pub fn from_zval(zval: ZVal) -> Self {
+        Self { zval }
+    }
+}
+
+impl From<ConstantValue> for Value {
+    fn from(constant: ConstantValue) -> Self {
+        Value {
+            zval: constant.into(),
+        }
+    }
+}
+
+macro_rules! implement_try_into {
+    ($type:ty, $fn:ident) => {
+        impl TryInto<$type> for Value {
+            type Error = Error;
+
+            fn try_into(self) -> Result<$type, Self::Error> {
+                self.zval.$fn()
+            }
+        }
+    };
+}
+
+implement_try_into!(i64, expect_long);
+implement_try_into!(f64, expect_double);
+implement_try_into!(bool, expect_bool);
+implement_try_into!((), expect_null);
+
+impl TryInto<String> for Value {
+    type Error = Error;
+
+    fn try_into(self) -> Result<String, Self::Error> {
+        Ok(self.zval.expect_z_str()?.to_str()?.to_string())
+    }
+}
+
+impl TryInto<Vec<String>> for Value {
+    type Error = Error;
+
+    fn try_into(mut self) -> Result<Vec<String>, Self::Error> {
+        self.zval
+            .expect_mut_z_arr()?
+            .iter()
+            .map(|v| Ok::<String, Error>(v.1.expect_z_str()?.to_str()?.to_string()))
+            .collect::<Result<Vec<String>, Error>>()
+    }
+}
+
+unsafe impl Sync for Value {}
+unsafe impl Send for Value {}
+
+#[derive(Clone)]
+pub enum ArrayKeyValue {
+    String(String),
+    Int(u64),
```

Comment:
should be `i64`, as PHP allows arrays to have negative value indexes, however, `crate::arrays::InsertKey` only supports `u64` for now. should be fixed.

Reply:
Oh, you can open a new PR to fix it.

Reply:
I can't find a way to pass i64/i32 to PHP in zend API, i have asked in r13, hopefully one of PHP internals can provide a way to do this, will PR once i figure it out :)